src/ProgramParser.test.js

```diff
@@ -13,66 +16,213 @@ test('Parse unsupported character', () => {
+    expect((new ProgramParser()).parse('D')).toStrictEqual({
+        program: [{block: 'left180'}],
+        highestLoopNumber: 0
+    });
 });
 
 test('Parse a', () => {
-    expect((new ProgramParser()).parse('a')).toStrictEqual([{block: 'right45'}]);
+    expect((new ProgramParser()).parse('a')).toStrictEqual({
+        program: [{block: 'right45'}],
+        highestLoopNumber: 0
+    });
 });
 
 test('Parse b', () => {
-    expect((new ProgramParser()).parse('b')).toStrictEqual([{block: 'right90'}]);
+    expect((new ProgramParser()).parse('b')).toStrictEqual({
+        program: [{block: 'right90'}],
+        highestLoopNumber: 0
+    });
 });
 
 test('Parse d', () => {
-    expect((new ProgramParser()).parse('d')).toStrictEqual([{block: 'right180'}]);
+    expect((new ProgramParser()).parse('d')).toStrictEqual({
+        program: [{block: 'right180'}],
+        highestLoopNumber: 0
+    });
+});
+
+test('Parse startLoop', () => {
+    expect((new ProgramParser()).parse('sA1sz')).toStrictEqual({
+        program: [
+            {
+                block: 'startLoop',
+                iterations: 1,
+                label: 'A'
+            },
+            {
+                block: 'endLoop',
+                label: 'A'
+            }
+        ],
+        highestLoopNumber: 1
+    });
+
+    expect((new ProgramParser()).parse('sAA12sz')).toStrictEqual({
+        program: [
+            {
+                block: 'startLoop',
+                iterations: 12,
+                label: 'AA'
+            },
+            {
+                block: 'endLoop',
+                label: 'AA'
+            }
+        ],
+        highestLoopNumber: 27
+    });
+
+    expect(() => {
+        (new ProgramParser()).parse('s');
+    }).toThrowError(/^Missing loop label$/);
+
+    expect(() => {
+        (new ProgramParser()).parse('ss');
+    }).toThrowError(/^Missing loop label$/);
+
+    expect(() => {
+        (new ProgramParser()).parse('s1');
+    }).toThrowError(/^Missing loop label$/);
+
+    expect(() => {
+        (new ProgramParser()).parse('sAAA');
+    }).toThrowError(/^Loop label too long: AAA$/);
+
+    expect(() => {
+        (new ProgramParser()).parse('sA1ssA1s');
+    }).toThrowError(/^Duplicate loop label: A$/);
+
+    expect(() => {
+        (new ProgramParser()).parse('sA');
+    }).toThrowError(/^Missing loop number of iterations$/);
+
+    expect(() => {
+        (new ProgramParser()).parse('sAs');
+    }).toThrowError(/^Missing loop number of iterations$/);
+
+    expect(() => {
+        (new ProgramParser()).parse('sA100s');
+    }).toThrowError(/^Loop has too many iterations: 100$/);
+
+    expect(() => {
+        (new ProgramParser()).parse('sA1');
+    }).toThrowError(/^Missing startLoop terminating 's'$/);
+
+    expect(() => {
+        (new ProgramParser()).parse('sA1x');
+    }).toThrowError(/^Missing startLoop terminating 's'$/);
+});
+
+test('Parse invalid loop', () => {
```

Comment:
Suggested rename to 'Parse unbalanced startLoop and endLoop'. 'Parse invalid loop' is too general as the tests above under 'Parse startLoop' also include invalid loops.